pyomo/gdp/disjunct.py

```diff
@@ -207,8 +244,16 @@ def __getstate__(self):
 
     def set_value(self, expr):
         for e in expr:
-            # The user gave us a proper Disjunct block
-            if hasattr(e, 'type') and e.type() == Disjunct:
+            # The user gave us a proper Disjunct block 
+            # [ESJ 06/21/2019] This is really an issue with the reclassifier,
+            # but in the case where you are iteratively adding to an
+            # IndexedDisjunct indexed by Any which has already been transformed,
+            # the new Disjuncts are parading as Blocks already. This catches
+            # them for who they are anyway.
+            # [ESJ 09/13/2019] We still need this. I didn't dig again, but I am
+            # tempted to trust my past self.
+            if isinstance(e, _DisjunctData) or isinstance(e, SimpleDisjunct):
+            #if hasattr(e, 'type') and e.type() == Disjunct:
```

Comment:
Why the change to using `isinstance()`? Also, you could just do `isinstance(e, (_DisjunctData, SimpleDisjunct))`.

Reply:
This is a hack around the fact that we reclassify disjuncts as blocks at the end of bigm. So you could have an algorithm where you generate a new disjunction in each iteration and retransform, like this:
```
model.firstTerm = Disjunct(Any)
model.secondTerm = Disjunct(Any)
model.disjunctionList = Disjunction(Any)
        
for i in range(2):
       model.firstTerm[i].cons = Constraint(expr=model.x == 2*i)
       model.secondTerm[i].cons = Constraint(expr=model.x >= i + 2)
       model.disjunctionList[i] = [model.firstTerm[i], model.secondTerm[i]]

       TransformationFactory('gdp.bigm').apply_to(model)
```
After the first iteration, `firstTerm` and `secondTerm` are reclassified as blocks. So in all later iterations, when you add to them in the first two lines of the loop, you add blocks. The third line will throw a ValueError with the hasattr version because `e.type()` is `Block`. However, they are still `_DisjunctData` instances, hence the change.

Reply:
Ok, that makes sense. You could condense the two `isinstance()` calls into one, though: https://stackoverflow.com/a/33311330/1557256

Reply:
Testing both `_DisjunctData` and `SimpleDisjunct` is redundant.